x-pack/plugins/ai_infra/product_doc_base/server/services/package_installer/package_installer.ts

```diff
@@ -0,0 +1,217 @@
+  private readonly artifactsFolder: string;
+  private readonly esClient: ElasticsearchClient;
+  private readonly productDocClient: ProductDocInstallClient;
+  private readonly endpointManager: InferenceEndpointManager;
+  private readonly artifactRepositoryUrl: string;
+  private readonly currentVersion: string;
+
+  constructor({
+    artifactsFolder,
+    logger,
+    esClient,
+    productDocClient,
+    endpointManager,
+    artifactRepositoryUrl,
+    kibanaVersion,
+  }: PackageInstallerOpts) {
+    this.esClient = esClient;
+    this.productDocClient = productDocClient;
+    this.artifactsFolder = artifactsFolder;
+    this.endpointManager = endpointManager;
+    this.artifactRepositoryUrl = artifactRepositoryUrl;
+    this.currentVersion = majorMinor(kibanaVersion);
+    this.log = logger;
+  }
+
+  /**
+   * Make sure that the currently installed doc packages are up to date.
+   * Will not upgrade products that are not already installed
+   */
+  async ensureUpToDate({}: {}) {
+    const repositoryVersions = await fetchArtifactVersions({
+      artifactRepositoryUrl: this.artifactRepositoryUrl,
+    });
+
+    const installStatuses = await this.productDocClient.getInstallationStatus();
+
+    const toUpdate: Array<{
+      productName: ProductName;
+      productVersion: string;
+    }> = [];
+    Object.entries(installStatuses).forEach(([productName, productState]) => {
+      if (productState.status === 'uninstalled') {
+        return;
+      }
+      const availableVersions = repositoryVersions[productName as ProductName];
+      if (!availableVersions || !availableVersions.length) {
+        return;
+      }
+      const selectedVersion = selectVersion(this.currentVersion, availableVersions);
+      if (productState.version !== selectedVersion) {
+        toUpdate.push({
+          productName: productName as ProductName,
+          productVersion: selectedVersion,
+        });
+      }
+    });
+
+    for (const { productName, productVersion } of toUpdate) {
+      await this.installPackage({
+        productName,
+        productVersion,
+      });
+    }
+  }
+
+  async installAll({}: {}) {
+    const repositoryVersions = await fetchArtifactVersions({
+      artifactRepositoryUrl: this.artifactRepositoryUrl,
+    });
+    const allProducts = Object.values(DocumentationProduct) as ProductName[];
+    for (const productName of allProducts) {
+      const availableVersions = repositoryVersions[productName];
+      if (!availableVersions || !availableVersions.length) {
+        this.log.warn(`No version found for product [${productName}]`);
+        continue;
+      }
+      const selectedVersion = selectVersion(this.currentVersion, availableVersions);
+
+      await this.installPackage({
+        productName,
+        productVersion: selectedVersion,
+      });
+    }
+  }
+
+  async installPackage({
+    productName,
+    productVersion,
+  }: {
+    productName: ProductName;
+    productVersion: string;
+  }) {
```

Comment:
The product doc install workflow logic 
- uninstall before install if present
- download the artifact
- unzip
- validate
- create the index 
- ingest content
- update the installation status

refer to individual steps for implementation details

Reply:
If this process fails somewhere in the middle, then we will end up with a set of product docs that are marked as `uninstalled`, which means we will never attempt to automatically update it again via `ensureUpToDate`. I don't want to over complicate the process, but is there a way to make this more resilient to failures?